cdci_data_analysis/analysis/email_helper.py

```diff
@@ -0,0 +1,220 @@
+import time as time_
+from email.mime.text import MIMEText
+from email.mime.multipart import MIMEMultipart
+import typing
+from ..analysis import tokenHelper
+import smtplib
+import ssl
+import os
+import re
+import glob
+from jinja2 import Environment, FileSystemLoader
+from flask import Markup
+
+from ..analysis.exceptions import BadRequest, MissingRequestParameter
+
+from datetime import datetime
+
+class EMailNotSent(BadRequest):
+    pass
+
+def timestamp2isot(timestamp_or_string: typing.Union[str, float]):
+    try:
+        timestamp_or_string = float(timestamp_or_string)
+    except ValueError:
+        pass
+
+    if isinstance(timestamp_or_string, float):
+        return datetime.fromtimestamp(float(timestamp_or_string)).strftime("%Y-%m-%d %H:%M:%S")
+        
+    return timestamp_or_string
+
+def textify_email(html):
+    text = re.search('<body>(.*?)</body>', html, re.S).group(1)
+
+    text = re.sub('<title>.*?</title>', '', text)
+
+    text = re.sub('<a href=(.*?)>(.*?)</a>', r'\2: \1', text)
+
+    return re.sub('<.*?>', '', text)
+   
+    #return Markup(text).striptags()
+
+def send_email(
+        config,
+        logger,
+        decoded_token,
+        job_id,
+        status="done",
+        instrument="",
+        product_type="",
+        time_request=None,
+        request_url="",
+        api_code=""):
+
+    # let's get the needed email template; 
+    # TODO: should get from pkgresources or so
+    env = Environment(loader=FileSystemLoader('%s/../flask_app/templates/' % os.path.dirname(__file__)))
+    env.filters['timestamp2isot'] = timestamp2isot
+
+    api_code_no_token = re.sub('"token": ".*?"', '"token": "<PLEASE-INSERT-YOUR-TOKEN-HERE>"', api_code)\
+                        .strip()\
+                        .replace("\n", "<br>\n")
+    
+    email_data = {
+        'oda_site': { 
+            #TODO: get from config
+            'site_name': 'University of Geneva',
+            'frontend_url': 'https://www.astro.unige.ch/mmoda', 
+            'contact': 'contact@odahub.io'
+        },
+        'request': {
+            'job_id': job_id,
+            'status': status,
+            'instrument': instrument,
+            'product_type': product_type,
+            'time_request': time_request,
+            'request_url': request_url,
+            'api_code_no_token': api_code_no_token,
+        }
+    }
+
+    template = env.get_template('email.html')
+    email_body_html = template.render(**email_data)
+    
+    email_subject = re.search("<title>(.*?)</title>", email_body_html).groups()[0]    
```

Comment:
Not important, but why not .group(1)?

Reply:
no particular reason. Is it not unfortunate groups are counted from 1?..